daemon/algod/api/server/auth.go

```diff
@@ -0,0 +1,168 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package server
+
+import (
+	"crypto/subtle"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/labstack/echo/v4"
+
+	"github.com/algorand/go-algorand/daemon/algod/api/server/lib/middlewares"
+	"github.com/algorand/go-algorand/logging"
+)
+
+const urlAuthFormatter = "/urlAuth/%s"
+const debugRouteName = "debug"
+
+// AuthRoutes define the desired mapping of authentication token keys to allowed routes.
+type AuthRoutes map[string]map[echo.Route]echo.HandlerFunc
+
+// Auth is the authentication layer manager. It creates a wrapper around each invoked call and provide authentication.
+type Auth struct {
+	log  logging.Logger
+	echo *echo.Echo
+}
+
+// MakeAuth create an authentication object
+func MakeAuth(log logging.Logger, e *echo.Echo) *Auth {
+	return &Auth{
+		log:  log,
+		echo: e,
+	}
+}
+
+// RegisterHandlers registers the given authentication routes
+func (auth *Auth) RegisterHandlers(authRoutes AuthRoutes) {
+	routeTokens := make(map[echo.Route][]string)
+
+	// for each unique route, find the list of tokens.
+	for token, routes := range authRoutes {
+		for route := range routes {
+			routeTokens[route] = append(routeTokens[route], token)
+		}
+	}
+	for _, routes := range authRoutes {
+		for route, routeFunction := range routes {
+			auth.echo.Add(route.Method, route.Path, makeAuthenticatedHandler(routeTokens[route], routeFunction, route).GetHandler())
+		}
+	}
+}
+
+// authenticatedHandler is a helper struct to "tunnel" the correct authntication scheme for
+// given function. ( no-auth, debug, regular auth )
+type authenticatedHandler struct {
+	tokens  [][]byte
+	handler echo.HandlerFunc
+	route   echo.Route
+}
+
+// makeAuthenticatedHandler creates the authentication handler.
+func makeAuthenticatedHandler(tokens []string, handler echo.HandlerFunc, route echo.Route) *authenticatedHandler {
+	authHandler := &authenticatedHandler{
+		handler: handler,
+		route:   route,
+	}
+	for _, token := range tokens {
+		authHandler.tokens = append(authHandler.tokens, []byte(token))
+	}
+
+	return authHandler
+}
+
+// GetHandler retrives the "correct" handler function wrapper for the underlaying handler.
+// in case of no-auth, it use the underlaying handler directly.
+func (h *authenticatedHandler) GetHandler() echo.HandlerFunc {
+	// if this entrypoint require no authentication, use that
+	// as this is the least restrictive.
+	for _, token := range h.tokens {
+		if len(token) == 0 {
+			// no authentication is needed
+			return h.handler
+		}
+	}
+
+	if h.route.Path == "/urlAuth/:token/debug/pprof/*" {
+		return h.DebugHandler
+	}
+	return h.Handler
+}
+
+// DebugHandler is the pprof debug handler, and used only for the `/urlAuth/:token/debug/pprof/*` path
```

Comment:
Do you know why the pprof endpoint needs to have the token provided as part of the path?
Why not use the same header as everything else and skip the special case?

Reply:
`go tool pprof` can auto fetch from a URL, but can't set headers